src/docker/ContextChangeCancelClient.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Disposable } from 'vscode';
+import { IActionContext, UserCancelledError } from 'vscode-azureextensionui';
+import { CancellationToken } from 'vscode-languageclient';
+import { CancellationPromiseSource, getCancelPromise, TimeoutPromiseSource } from '../utils/promiseUtils';
+import { ContextManager } from './ContextManager';
+import { DockerContext } from './Contexts';
+
+export abstract class ContextChangeCancelClient implements Disposable {
+    private readonly contextChangedDisposable: Disposable;
+    protected contextChangeCps: CancellationPromiseSource;
+
+    public constructor(private readonly contextManager: ContextManager) {
+        this.contextChangeCps = new CancellationPromiseSource();
+        this.contextChangedDisposable = this.contextManager.onContextChanged((currentContext: DockerContext) => this.onContextChange(currentContext))
+    }
+
+    protected onContextChange(currentContext: DockerContext): void {
+        this.contextChangeCps.cancel();
+        this.contextChangeCps.dispose();
+        this.contextChangeCps = new CancellationPromiseSource();
+    }
+
+    public dispose(): void {
+        this.contextChangeCps.cancel();
+        this.contextChangeCps.dispose();
+        this.contextChangedDisposable.dispose();
+    }
+
+    protected async withTimeoutAndCancellations<T>(context: IActionContext, callPromise: () => Promise<T>, timeout: number, token?: CancellationToken): Promise<T> {
+        const tps = new TimeoutPromiseSource(timeout);
+        const evt = tps.onTimeout(() => {
+            context.errorHandling.suppressReportIssue = true;
+        });
+
+        try {
+            const promises: Promise<T>[] = [tps.promise, this.contextChangeCps.promise, callPromise()];

```

Comment:
Isn't it nice that `Promise<never>` can act as `Promise<T>`? ☺️

Reply:
I'm glad, since it was the only way to implement this that the transpiler was happy with!